### PR TITLE
Feat/derived columns editor

### PR DESCRIPTION
**Pull request by macintoshpie**
#### Any background context you want to provide?
SEED is adding ability for users to create custom columns derived from other column data.

#### What's this PR do?
Adds views and frontend for creating, editing, and deleting derived columns.

What this PR does NOT do:
We have discussed adding a "preview" of the derived column result for some inventory but that is NOT part of this PR. It will likely be added at a later time.

We have discussed adding a "warning" for unit aware columns. This is NOT included in this PR, I need to chat with someone more familiar with how to determine whether a column is unit aware. It will be added at a later time.

#### How should this be manually tested?
Go to your organization page and click the "Derived Columns" tab. You should be able to create, update, and delete derived column data. You should get helpful errors when things are invalid / incorrect.

Some notes:
- you should only be able to select columns for your currently active inventory type
- no duplicate parameter names or source columns are allowed
- parameter names must be a [valid Python identifier](https://docs.python.org/3/reference/lexical_analysis.html#identifiers)
- you must use all parameters in the expression
- you can't have undeclared parameters in the expression
- you must have at least one parameter (delete button should disappear)
- you can't edit the inventory type of an existing derived column
- the "create" or "save" button should generally be disabled if there are any active errors (except for those returned by the server, which are located below the expression and below the `Error(s):` title

#### What are the relevant tickets?
#1642 

#### Screenshots (if appropriate)
Editing existing column example
Screen Shot 2021-05-25 at 11 20 02 AM

Example of different errors detected on frontend
Screen Shot 2021-05-25 at 11 22 50 AM

Example of errors detected on backend
Screen Shot 2021-05-25 at 11 23 38 AM

### CODE REVIEW

File: seed/static/seed/partials/derived_columns_editor.html [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+                </button>
+                <span></span>
+                <span></span>
+
+                <!-- expression -->
+                <h3>Expression</h3>
+                <span></span>
+                <span></span>
+                <div style="grid-column: 1 / 3">
+                    <input
```
Comment (adrian-lara) [inline seed/static/seed/partials/derived_columns_editor.html:116]: Probably a note for a future change: Related to creating an expression, it'd be great to see notes/instructions (maybe something that's collapsible?) showing the different operations and the syntax for each of them .

File: seed/static/seed/js/controllers/derived_columns_editor_controller.js [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+        // name is invalid if it doesn't follow Python variable naming conventions
+        // or if it has the same name as another parameter
+        this_param.errors.invalid_parameter_name = !/^[_a-zA-Z][_a-zA-Z0-9]*$/.test(this_param.parameter_name)
+        this_param.errors.duplicate_parameter_name = $scope.parameters.some((param, idx) => {
+          return param.parameter_name == this_param.parameter_name &&
+            idx != parameter_index
+        })
+      }
+
+      // updates parameter errors according to source column validity
```
Comment (adrian-lara) [inline seed/static/seed/js/controllers/derived_columns_editor_controller.js:120]:
```suggestion
      // updates parameter errors according to source column validity and check for duplicate
```

**Review by adrian-lara — APPROVED** [review]
This looks great! A few small changes (maybe?), but ideally, we get this in front of others to test sooner than later. I monkey tested some but getting extra eyes for more testing and other feedback would be really useful.

File: seed/static/seed/partials/derived_columns_editor.html (hunk 1 above)
Comment (macintoshpie) [reply seed/static/seed/partials/derived_columns_editor.html:116]: I totally agree, will add a collapsable help section later 👍

File: seed/views/v3/derived_columns.py [hunk 3]
```diff
@@ -4,20 +4,39 @@
...
+    class Meta:
+        model = DerivedColumn
+        fields = ['name', 'expression', 'inventory_type']
+
+    def create(self, validated_data):
+        return DerivedColumn.objects.create(
+            name=validated_data['name'],
+            expression=validated_data['expression'],
+            inventory_type=validated_data['inventory_type'],
+        )
```
[the remark this answers is not in the document]
Comment (macintoshpie) [reply seed/views/v3/derived_columns.py]: Whoops! Good catch, I ended up moving everything to the same serializer so I will delete this

File: seed/static/seed/js/services/derived_columns_service.js [hunk 4]
```diff
@@ -18,5 +18,47 @@ angular.module('BE.seed.service.derived_columns', []).factory('derived_columns_s
...
+      return $http({
+        url: `/api/v3/derived_columns/${derived_column_id}`,
+        method: 'GET',
+        params: { organization_id }
+      }).then(function (response) {
+        return response.data;
+      });
+    };
+
+    derived_columns_factory.create_derived_column = function(organization_id, { name, expression, inventory_type, parameters }) {
```
[the remark this answers is not in the document]
Comment (macintoshpie) [reply seed/static/seed/js/services/derived_columns_service.js:31]: 👁️ 👅 👁️

File: seed/static/seed/partials/derived_columns_editor.html [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+                        <span ng-repeat="(field_name, messages) in errors_from_server">
+                            <br>
+                            {$ field_name $}:
+                            <ul>
+                                <li ng-repeat="message in messages">{$ message $}</li>
+                            </ul>
+                        </span>
+                    </div>
+                </div>
+                <div>
```
[the remark this answers is not in the document]
Comment (macintoshpie) [reply seed/static/seed/partials/derived_columns_editor.html:134]: Hmm maybe I'll think about that as I update the page, there's still a bit that needs to be added

**Comment by RDmitchell** [comment]
instance: dev1
SHA: 
Org: LBNL 200

The way the UI is working right now is confusing. 

As soon as I started to define a parameter name, ie, "siteEnergy", the program put up the error message that the parameter hadn't been referenced in the expression -- this is confusing, and at first I thought I had to define it as $siteEnergy, but then I got the message that the parameter name was invalid. 

I think the message that the parameter is not used in the Expression shouldn't show up until the expression has been defined using the parameters. 

Maybe adding a "test" button to the Expression section, after the expression has been defined, would be the place to put up the appropriate error messages. It would be nice to have an example of the derived column calculation to make sure that what was defined is what the user wanted. 

image

**Comment by macintoshpie** [comment]
@RDmitchell I agree, adding a preview / test button (like [here](https://github.com/SEED-platform/seed/issues/1642#issuecomment-815114173)) would be ideal. We will try our best to get it included before the release.

I also think your idea of showing errors once the user clicks "test" is nice, I will include it in the future additions

**Pull request merged**